### PR TITLE
introduce soft delete image endpoint

### PR DESCRIPTION
**Pull request by abdelrahmansd**
## What does this change?
This implements a new endpoint: Soft Delete images

## How can success be measured?
Given `DELETE     media-api/soft-delete/images/:id` endpoint
and the endpoint hit with proper image-id
Then `softDeletedMetadata` fields should be added in the root of the document 
`                      "softDeletedMetadata": {
                        "deleteTime": "2021-06-17T03:52:37.084+02:00",
                        "deletedBy": "johndoe@example.com"
                    }`

## Tested?
- [x] locally by committer
- [ ] locally by Guardian reviewer
- [ ] on the Guardian's TEST environment

### CODE REVIEW

**Comment by twrichards** [comment]
LGTM so far - will wait for `deletedBy` before doing a final review

File: metadata-editor/app/controllers/EditsController.scala
Comment (twrichards) [inline metadata-editor/app/controllers/EditsController.scala:218]: it looks like we're discarding the output of this command and mapping on the result of setting the deletedBy only, is there a way we could combine them into a single command to set both fields?

**Review by twrichards — COMMENTED** [review]
it all looks nice and clean, I will test in our TEST environment tomorrow morning. One important comment on the code in the meantime though.

File: metadata-editor/app/controllers/EditsController.scala
Comment (abdelrahmansd) [reply metadata-editor/app/controllers/EditsController.scala:218]: yes we can use `jsonAdd` to add both
will update the PR now

File: metadata-editor/app/controllers/EditsController.scala
Comment (twrichards) [reply metadata-editor/app/controllers/EditsController.scala:219]: if however you don't like the _varargs_ part of my other suggestion (I don't have a strong preference), then please can we neaten up the call signature to make clear the two inputs, something like...
```suggestion
    editsStore.stringListSet(
       id, 
       Map(
          Edits.DeletedAt -> JsString(DateTime.now.toString),
          Edits.DeletedBy -> JsString(req.user.accessor.identity)
        )
      )
      .map(publish(id, UpdateImageUserMetadata))
```

**Review by twrichards — COMMENTED** [review]
thanks for updating to make the two field updates in a single operation

couple of style suggestions, but otherwise looks good. I'm going to test in our TEST environment now.

**Comment by twrichards** [comment]
tested in `TEST`, see the metadata after hitting the endpoint 
image

File: metadata-editor/app/controllers/EditsController.scala
Comment (twrichards) [reply metadata-editor/app/controllers/EditsController.scala]: done in https://github.com/guardian/grid/pull/3353/commits/b4c9fc9cd92e0267dfcee64540a636036b01b151

File: common-lib/src/main/scala/com/gu/mediaservice/lib/aws/DynamoDB.scala
Comment (twrichards) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/aws/DynamoDB.scala]: done in https://github.com/guardian/grid/pull/3353/commits/b4c9fc9cd92e0267dfcee64540a636036b01b151

**Review by twrichards — APPROVED** [review]
Now I've tested this, I'm happy to merge tomorrow. Nice work 🙌 

We really ought to add this endpoint to the 'hypermedia' responses on the GET endpoints (i.e. when reading this information it provides the url for all the actions, in this case `soft-delete` so the client doesn't need to build the URL itself). Will leave it up to you whether you want to make that part of this PR or a follow-up? (You'll need it when you move onto the client-side portion)

**Comment by twrichards** [comment]
we're observing that `thrall` isn't booting up on this branch, so I'm going to hold-off from merging for now. The problem relates to adding these fields to ES, you got any thoughts @akash1810 ?
image

**Comment by paperboyo** [comment]
Inability to change the existing mappings is one of the reasons we are doing the reingestion work. But I thought adding a new one is not a problem. No idea if [this script](https://github.com/guardian/grid/tree/main/scripts#updatemapping) is actually how you do it, though…

**Comment by abdelrahmansd** [comment]
@twrichards it will require mapping update because `userMetadataMapping` set to strict

File: common-lib/src/main/scala/com/gu/mediaservice/lib/elasticsearch/Mappings.scala
Comment (akash1810) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/elasticsearch/Mappings.scala:193]: I don't think this is the correct location for these fields as it suggests this is the time the user metadata was deleted, rather than the image itself, which is what we want to represent. A better location would be at the [root of the document as siblings to `uploadTime` and `uploadedBy`](https://github.com/guardian/grid/blob/c7ccb0ab6603209aaa816188e2547fb7394d90af/common-lib/src/main/scala/com/gu/mediaservice/lib/elasticsearch/Mappings.scala#L53-L54).

We may also want to call it `deletedTime` for consistency?

**Comment by twrichards** [comment]
> Inability to change the existing mappings is one of the reasons we are doing the reingestion work. But I thought adding a new one is not a problem. No idea if [this script](https://github.com/guardian/grid/tree/main/scripts#updatemapping) is actually how you do it, though…

Running this script did the trick, thanks @paperboyo (and @akash1810 for mobbing on running it). I think the point that @akash1810  raises about placement of these fields is important though, so will hold off merging until for now https://github.com/guardian/grid/pull/3353#discussion_r652569008

File: common-lib/src/main/scala/com/gu/mediaservice/lib/elasticsearch/Mappings.scala
Comment (abdelrahmansd) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/elasticsearch/Mappings.scala:193]: @akash1810 yes it will be making more sense 
will change the name and move it to the `Image` model

Comment (abdelrahmansd) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/elasticsearch/Mappings.scala]: @akash1810 I couldn't add `deleteTime` and `deletedBy` at the root of the document  because we hit  function parameters maximum number `implementation restricts functions to 22 parameters`
so I added them into a nested object `softDeletedMetadata`

Comment (akash1810) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/elasticsearch/Mappings.scala]: Ah, Scala, such a wonderful language 😉 .

File: media-api/app/controllers/MediaApi.scala
Comment (twrichards) [inline media-api/app/controllers/MediaApi.scala:238]: no need for val as case class name describes what it is, named params on the SoftDeletedMetadata args and changed line breaks & indentation to make things clearer  
```suggestion
        messageSender.publish(
          UpdateMessage(
            subject = SoftDeleteImage, 
            id = Some(id),
            softDeletedMetadata = Some(SoftDeletedMetadata(
              deleteTime = DateTime.now(DateTimeZone.UTC),
              deletedBy = request.user.accessor.identity
            )
          )
        )
```

**Review by twrichards — COMMENTED** [review]
Just testing now, one minor comment in the meantime - otherwise looks good 👍 

Also if you don't mind, please could you squash the commits (I would happily do it at merge time but don't want to mess up BBC when you pull the changes back through to the bbc repo)

File: media-api/app/controllers/MediaApi.scala
Comment (abdelrahmansd) [reply media-api/app/controllers/MediaApi.scala:238]: @twrichards will update and squash the PR now

**Review by twrichards — APPROVED** [review]
re-tested 🎉 
image

Amazing work and thanks for your patience though the various suggestions/iterations.

Will merge once squashed :)

**Pull request merged**
